### ISSUE TITLE
Install algorithm step 14 should clear waiting worker before updating state to redundant

### ISSUE THREAD

**Issue by wanderview** [slightlyoff/ServiceWorker#851]
The Install algorithm has this currently

```
14) If registration’s waiting worker is not null, then:
  1) Terminate registration’s waiting worker.
  2) Run the Update State algorithm passing registration’s waiting worker and
       redundant as the arguments.
  3) The user agent may abort in-flight requests triggered by registration’s waiting worker. 
15) Set registration’s waiting worker to registration’s installing worker. 
```

It seems the `statechange` event triggered by 14.2 Update State might see the redundant worker still set as `registration.waiting`.  Most other instances of Update State change the registration attribute first.

As a side note, should the `statechange` event see the new service worker in registration.waiting or a null .waiting?

**Comment by wanderview** [comment]
Maybe the steps should be something like

1. Set old waiting worker to registration's waiting worker.
2. Set registration's waiting worker to registration's installing worker.
3. Clear registration's installing worker.
4. If old waiting worker is not null, then:
  1. Terminate old waiting worker.
  2. Run the Update State algorithm to redundant
  3. The user-agent may terminate...

**Comment by wanderview** [comment]
Or does this not matter because the statechange event dispatch uses "queue a task"?

**Comment by wanderview** [comment]
So, I think since we "queue a task" to fire the statechange event, then this all works fine for DOM objects on the main thread.  I think there is a problem for DOM objects on workers, though.

ServiceWorker DOM objects exposed on a worker thread are spec'd to access attributes like `.installing` like this:

> Let waitingWorker be the ServiceWorker object that represents the service worker registration’s waiting worker.

This seems to have immediate and synchronous access to the `registration's waiting worker`.

When we queue a task to fire the statechange event on the worker thread, however, it may fire before the `registration's waiting worker` is updated.  So worker threads may see an inconsistent view of ServiceWorker and ServiceWorkerRegistration DOM object state.

Now, if the spec said to queue a task to update the ServiceWorkerRegistration DOM object on the worker thread, then we could rely on the "queue a task" language to order things correctly for the worker thread.

Also, all of this would be much easier to understand if we had a note for each Update State clarifying what we expect the ServiceWorkerRegistration and ServiceWorker info objects to look like when the event happens.  Its hard to ensure both window and worker threads see the same state currently.

**Comment by jungkees** [comment]
Well, let me try to clarify what we want to make clear:

> It seems the statechange event triggered by 14.2 Update State might see the redundant worker still set as registration.waiting. Most other instances of Update State change the registration attribute first.

The order of Invoking Update State and changing the registration's attribute varies case by case depending on which `ServiceWorker` objects in the client contexts should be notified the change event.

Note that `ServiceWorker` object captured in a client context (either main thread or worker) has an associated service worker determined at the time of the execution of the getters (`.installing/.waiting/.active/.controller`). And those associated service workers' state change during the installation process. What Update State does is finding all the `ServiceWorker` objects associated with the given service worker and queue a task to fire the statechange events to those client contexts. For example, the Install step 17 invokes Update State with registrations.waiting after setting its value to registration.installing in the step 15. Then from the script surface, all the `ServiceWorker` objects captured by `registration.installing` before these Install step reached will be getting dispatched the `statechange` event, its state having changed to `redundant`.

In the same way, what the Install step 14 does in the OP is to find all the `ServiceWorker` objects associated with the current registration.waiting (before nullified) and queue a task to those clients to notify now it's redundant.

> As a side note, should the statechange event see the new service worker in registration.waiting or a null .waiting?

I think registration.waiting inside the event listener of the statechange event should be a new service worker (the installing worker promoted to waiting worker.) So then we may need more precise description that the Install step 15 should be run before the event listener of the statechange event in the main thread is invoked?

**Comment by wanderview** [comment]
I think we need to think about how it works for `statechange` event listeners in worker threads as well.  We can assume tasks run in order between the main thread and worker threads.  So we probably need some "wait" step.

**Comment by wanderview** [comment]
If you would like I can figure out what I think makes sense and try to make a PR in the next week or two.

**Comment by jungkees** [comment]
> I think we need to think about how it works for statechange event listeners in worker threads as well. We can assume tasks run in order between the main thread and worker threads.

I'm not sure if I understood your concern here correctly. Update State is using either a main thread's event loop or a worker's event loop exclusively depending on where the target `ServiceWorker` object is stored: `The task must use serviceWorker’s relevant settings object’s responsible event loop and the DOM manipulation task source.` <- This step covers both cases. Am I still misunderstanding your point?

> If you would like I can figure out what I think makes sense and try to make a PR in the next week or two.

Sure. Your contribution and help are always welcome! Thanks.

**Comment by jungkees** [comment]
> Update State is using either a main thread's event loop or a worker's event loop exclusively depending on where the target `ServiceWorker` object is stored

Well, let me clarify it further as I think it can be a bit misleading. So, here's an example:
If (1) a page has `ServiceWorker` object 1, (2) a worker has `ServiceWorker` object 2, and (3)  a serviceworker has `ServiceWorker` object 3 to the same service worker gotten from the getters, Update State will queue a task to (1)'s event loop, (2)'s event loop, and (3)'s loop separately.

**Comment by wanderview** [comment]
> Well, let me clarify it further as I think it can be a bit misleading. So, here's an example:
If (1) a page has ServiceWorker object 1, (2) a worker has ServiceWorker object 2, and (3) a serviceworker has ServiceWorker object 3 to the same service worker gotten from the getters, Update State will queue a task to (1)'s event loop, (2)'s event loop, and (3)'s loop separately.

Sure, but consider what `registration.installing`, `.waiting`, and `.active` will say in those 3 different event handlers.

On the main thread we are guaranteed that the registration attributes will not change while the event is being handled.  In addition, we can accurately predict the state of the registration attributes by examining when tasks are queued.

Event loops running on separate threads, though, may handle the event simultaneous with the main thread changing the registration attributes.  In addition, its extremely racy as to what state the registration attributes will be in when the event handler is fired.

Another way to handle this would be to be explicit about how we reflect registration attributes into the worker event loop context.  The current spec says they are directly read which implies shared state across threads.  Instead, the spec could indicate the worker DOM objects get tasks queued to update the worker thread exposed attribute.  Then we could reason about how these attributes change consistently with when the state change events fire.

**Comment by jakearchibald** [comment]
F2F: Update attributes before firing events

**Comment by slightlyoff** [comment]
Here's Clear Site Data: https://w3c.github.io/webappsec-clear-site-data/

**Comment by wanderview** [comment]
I believe we will fix this in gecko bug https://bugzilla.mozilla.org/show_bug.cgi?id=1257977

**Comment by jungkees** [comment]
Addressed: 3266af01ca3ab891d330a86ba3182bc5dae6c9b2

**Comment by wanderview** [comment]
@jungkees Overall these changes look great!

I think in this case, though, we still need to fix up Install Algorithm step 14.  As it stands now when you fire the state change event for 14.2 the registration.waiting attribute will still point to the worker being set to redundant.

I think we either need to add a:

> Run the Update Registration state algorithm passing registration, "waiting", and null as the arguments.

Or remember the old waiting worker in a variable and then run Update Worker State after step 16.  This would make the redundant worker state change event also see the new worker in the waiting spot.

I think the second option here is slightly nicer for devs.

Does that make sense?  Thanks again!

**Comment by jungkees** [comment]
@wanderview, thank you so much for this feedback! You're right. I went for the 2nd option: c9f8a9fbd0a02ff2c48e9dcbbfe5ff34d26a9a6b Note also that I changed Update Registration State and Update Worker State to use microtasks rather than normal tasks. Otherwise, register()'s promise fulfillment callback (invoked by Install step 7) won't see the changed state queued by step 4 and step 5. Please take a look.

**Comment by wanderview** [comment]
I'm not sure its correct to queue a microtask across thread boundaries like that, is it?

I think it would be simpler to "queue a task to run Resolve Job Promise".  We're already in a separate task from the scheduling of the job, so the extra event loop turn should be non-observable to the script.  It would let use keep using normal tasks for other things, though.

One other nit.  I think Step 19 should only run the Update Worker State algorithm for redundantWorker if redundantWorker is not null.

**Comment by jungkees** [comment]
> One other nit. I think Step 19 should only run the Update Worker State algorithm for redundantWorker if redundantWorker is not null.

Addressed: 50f6d3bb8e9923e8c5eaf2978236901a48296957

> I'm not sure its correct to queue a microtask across thread boundaries like that, is it?

I think queuing a microtask from a parallel algorithm thread to a main thread makes sense. To me, "queue a task to run Resolve Job Promise" rather sounds a bit odd as the reason why promise settlement callback uses a microtask is to run the callback immediately before running the next task in the queue. Using a microtask to change the registration/service worker state seems like a good fit where the changed states can be immediately observed by subsequent tasks (and microtasks).

**Comment by jungkees** [comment]
Added a missed bit: b8480546e43db192791ea3674c4585f0a91c8c84

@annevk, need your help. Does it make sense to use a microtask for the state change of a registration and its service workers here? My concern was that if it used a normal task, the register() method's promise fulfillment callback which is followed by those steps wouldn't observe the changes. (i.e. registration.installing in the callback would be null.) Or just queuing a task to resolve the promise as @wanderview suggested would make better sense?

**Comment by wanderview** [comment]
Part of my issue with doing these as microtasks is now we are firing the events as microtasks as well.

Also, one of the goals of the task queueing change was to be able to reason about the order of events and changes on the different threads.  If some things happen via tasks and others via microtasks it gets very difficult to figure out what happens when.  As I understand it, queuing these as microtasks can make them fire before other queued tasks.

This changes a lot of timing in the spec just to accommodate one promise getting resolved.  It just seems safer to me to resolve the promise in a task.  Script won't be able to tell anyway since we're already in a task.  If implementations want to optimize this out in an equivalent way, thats fine.  But they would have to ensure the sane ordering of events.

**Comment by annevk** [comment]
I think @martinthomson had a security problem with queuing microtasks from "in parallel" algorithms, but I cannot find context.

@jungkees presumably the `register()` method's promise is also resolved from some task, right? If that is queued later, and they are on the same queue, there would not be a problem here.

**Comment by jungkees** [comment]
> @jungkees presumably the register() method's promise is also resolved from some task, right? If that is queued later, and they are on the same queue, there would not be a problem here.

This is presumably true under the assumption that a task is queued for "resolve" step. I haven't explicitly queued a task for that here. I'm wondering if "resolve" from a thread running in parallel implicitly does queue a task or not. (@domenic, can you help?) I presume it does not as it just synchronously invokes ES resolve function: https://www.w3.org/2001/tag/doc/promises-guide#shorthand-manipulating I think I found some similar case in HTML: https://html.spec.whatwg.org/#dom-createimagebitmap (see the SVG image element case there.) So, explicitly queuing a task to resolve a promise is required from a thread running in parallel? Or can I assume it's impliclity done?

Addressed the comment anyway (now it uses a normal task): 099deb9509eaf7348efb98ef6aaa6ac17d403d43

**Comment by martinthomson** [comment]
@annevk, the concern is only the scheduling of microtasks on a currently running (macro) task from an external thread.  That creates a potential violation of run-to-completion semantics.

**Comment by domenic** [comment]
Resolve is meant to queue a task if running in parallel: https://www.w3.org/2001/tag/doc/promises-guide#shorthand-manipulating

> If the algorithm using these shorthands is running in parallel, the shorthands queue a task on p’s relevant settings object’s responsible event loop to call the stored function.

**Comment by martinthomson** [comment]
OK, that is how I had hoped it would work.  My concerns aren't relevant then.

**Comment by annevk** [comment]
I think we should do two things:

1. Require "in parallel" algorithms to queue tasks when they want to resolve promises. E.g., the promises guide doesn't specify an "event source" and that can be problematic if ordering is important as it is here.
2. Make it clear that "in parallel" algorithms can only ever queue tasks, and not queue microtasks. The latter is rather unpredictable as the microtask could end up in whatever task happens to be running, which is quite strange, a security issue, not what is implemented, etc.

**Comment by jungkees** [comment]
> Resolve is meant to queue a task if running in parallel

Yeah, I had missed the paragraph right below.

> 1. Require "in parallel" algorithms to queue tasks when they want to resolve promises. E.g., the promises guide doesn't specify an "event source" and that can be problematic if ordering is important as it is here.

Can't we assume or specify (in the promises guide) the event source for the promise settlement would be the DOM manipulation task source? I like the way the promises guide does queue a task so the specs use it don't have to do it themselves everywhere. I guess the order between other task sources and the DOM manipulation task source won't matter much.

**Comment by annevk** [comment]
1. The promises guide is just a placeholder until that language lands in IDL. It's a rather weird setup the way we have it now.
2. It gets hairy when you have to queue a task and resolve the promise from there (does that queue another task?).
3. I think it's much better to be explicit about this than relying on this kind of indirection.

**Comment by jungkees** [comment]
> The promises guide is just a placeholder until that language lands in IDL. It's a rather weird setup the way we have it now.

I agree to the direction but think the promises guide is the reference that other promise-using specs should rely on in the mean time.

> It gets hairy when you have to queue a task and resolve the promise from there (does that queue another task?).

If you queued a task to settle a promise, then wouldn't it be running in a main thread already (in which case https://github.com/slightlyoff/ServiceWorker/issues/851#issuecomment-215283809 says it doesn't queue another task)?

Wouldn't it make sense that IDL specifies the following behavior?
> If the algorithm using these shorthands is running in parallel, the shorthands queue a task on p’s relevant settings object’s responsible event loop to call the stored function **using the DOM manipulation task source**.

**Comment by wanderview** [comment]
If we want to continue discussing possibly using microtask here, can we move it to a new issue?

**Comment by jungkees** [comment]
@wanderview I already pushed https://github.com/slightlyoff/ServiceWorker/commit/099deb9509eaf7348efb98ef6aaa6ac17d403d43 that made we use normal tasks again. The remaining issue here is whether we can rely on the promises guide (https://github.com/slightlyoff/ServiceWorker/issues/851#issuecomment-215283809) for the timing being until we have normative Web IDL reference on this. I think we can just close this if @annevk agrees.

**Comment by annevk** [comment]
FWIW, I don't think IDL will adopt this language as-is. You need to define the task source and I'm not really sure why the DOM manipulation task source would make sense here or why it would make sense for all promises. It certainly wouldn't for Fetch.

**Comment by jungkees** [comment]
I think settling a promise to give a platform-minted `ServiceWorkerRegistration` object to the main thread using the DOM manipulation task source makes sense. But I agree not all the promise would do so as you pointed with Fetch. I'll make it queue a task explicitly.

**Comment by jungkees** [comment]
@annevk @domenic Is a [promise](http://www.ecma-international.org/ecma-262/6.0/#sec-promise-objects) a [platform object](https://heycam.github.io/webidl/#dfn-platform-object)? According to [the definition in the Web IDL](https://heycam.github.io/webidl/#dfn-platform-object), it seems it's not. I'm curious if I can do "a promise _p_'s relevant settings object’s responsible event loop" assuming _p_ has been returned to the script surface.

**Comment by domenic** [comment]
Promise is not a platform object. You can probably just use the current settings object.

**Comment by jungkees** [comment]
Referencing the current settings object of the promise from within an algorithm running in parallel makes sense? I'm trying to get the event loop to queue a task to resolve that promise.

**Comment by domenic** [comment]
Oh, no, sorry, it does not. The current settings object is not present during algorithms running in parallel. You usually save it before going parallel, from the caller API.

(There is only one current settings object, not one for each object, BTW.)

**Comment by jungkees** [comment]
I see.. It seems the promises guide was doing it with the relevant settings object as mentioned in https://github.com/slightlyoff/ServiceWorker/issues/851#issuecomment-215283809

Hope that makes sense here too. Or do you have any words coming to mind?

**Comment by jungkees** [comment]
Extending the [Relevant section](https://html.spec.whatwg.org/#relevant-settings-object) to additionally define the relevant settings object for a promise can be an option?

**Comment by domenic** [comment]
I see. That's not great, as in general the relevant settings object is not defined. As @annevk says this needs to be better defined. You should ideally save a settings object from earlier in an algorithm.

**Comment by jungkees** [comment]
We actually had a handle to the settings object, _job_'s _client_. Addressed the issue: ff9e42562b0a60ddbad947f5f308122ca1cc8a42

Closing.
